source/patterns/@aws-solutions-constructs/aws-openapigateway-lambda/lib/index.ts

```diff
@@ -0,0 +1,205 @@
+import { RestApiBaseProps } from 'aws-cdk-lib/aws-apigateway';
+import { Asset } from 'aws-cdk-lib/aws-s3-assets';
+import { overrideProps } from '@aws-solutions-constructs/core';
+
+/**
+ * The ApiIntegration interface is used to correlate a user-specified id with either a existing lambda function or set of lambda props.
+ *
+ * See the 'Overview of how the OpenAPI file transformation works' section of the README.md for more details on its usage.
+ */
+export interface ApiIntegration {
+  /**
+   * Id of the ApiIntegration, used to correlate this lambda function to the api integration in the open api definition.
+   */
+  readonly id: string;
+  /**
+   * Existing instance of Lambda Function object. Providing both this and `lambdaFunctionProps` will cause an error.
+   */
+  readonly existingLambdaObj?: lambda.Function;
+  /**
+   * User provided props to override the default props for the Lambda function. Providing both this and `existingLambdaObj` will cause an error.
+   */
+  readonly lambdaFunctionProps?: lambda.FunctionProps;
+}
+
+/**
+ * Helper object to map an ApiIntegration id to its resolved lambda.Function.
+ */
+export interface ApiLambdaFunction {
+  /**
+   * Id of the ApiIntegration, used to correlate this lambda function to the api integration in the open api definition.
+   */
+  readonly id: string;
+  /**
+   * The instantiated lambda.Function.
+   */
+  readonly lambdaFunction: lambda.Function;
+}
+
+export interface OpenApiGatewayToLambdaProps {
+  /**
+   * S3 Bucket where the open-api spec file is located. When specifying this property, apiDefinitionKey must also be specified.
+   */
+  readonly apiDefinitionBucket?: s3.IBucket;
+  /**
+   * S3 Object name of the open-api spec file. When specifying this property, apiDefinitionBucket must also be specified.
+   */
+  readonly apiDefinitionKey?: string;
+  /**
+   * Local file asset of the open-api spec file.
+   */
+  readonly apiDefinitionAsset?: Asset;
+  /**
+   * One or more key-value pairs that contain an id for the api integration
+   * and either an existing lambda function or an instance of the LambdaProps.
+   *
+   * Example:
+   * const apiIntegrations: ApiIntegration[] = [
+   *   {
+   *     id: 'MessagesHandler',
+   *     lambdaFunctionProps: {
+   *       runtime: lambda.Runtime.NODEJS_16_X,
+   *       handler: 'index.handler',
+   *       code: lambda.Code.fromAsset(`${__dirname}/messages-lambda`),
+   *     }
+   *   },
+   *   {
+   *     id: 'PhotosHandler',
+   *     existingLambdaObj: new lambda.Function(this, 'PhotosLambda', {
+   *       runtime: lambda.Runtime.NODEJS_16_X,
+   *       handler: 'index.handler',
+   *       code: lambda.Code.fromAsset(`${__dirname}/photos-lambda`),
+   *     })
+   *   }
+   * ]
+   */
+  readonly apiIntegrations: ApiIntegration[];
+  /**
+   * Optional user-provided props to override the default props for the API.
+   *
+   * @default - Default props are used.
+   */
+  readonly apiGatewayProps?: RestApiBaseProps;
```

Comment:
ok so here is the thing. SpecRestApiProps is exactly RestApiBasProps + the ApiDefinition. I did this intentionally because:
1. we need to be able to transform the incoming template, but SpecRestApi doesn't expose it externally to us, so instead we have properties for the user to specify the ApiDefinition in the same way that they would in SpecRestAPi
2. using RestApiBaseProps means the user can't sneak a ApiDefinition into the rest api props and get unexpected behavior.